### PR TITLE
feat(tuner): add miner for session dataset

### PR DESCRIPTION
**Pull request by bwanglzu**
This PR extends the current `Miner` with `Session Miners`, generate all possible tuples/triplets w.r.t each session (anchor).

### CODE REVIEW

File: finetuner/tuner/pytorch/miner.py [hunk 1]
```diff
@@ -63,15 +63,21 @@ def mine(
...
         :return: a pair of label indices and their label as tuple.
         """
-        rv = []
         assert len(embeddings) == len(labels)
+        rv = []
+        labels_with_index = [item + (index,) for index, item in enumerate(labels)]
+        sorted_labels_with_index = sorted(labels_with_index, key=lambda x: x[0])
         for session, group in groupby(
-            sorted(labels, key=lambda x: x[0]), lambda x: x[0]
+            sorted(sorted_labels_with_index, key=lambda x: x[0]), lambda x: x[0]
```
Comment (JoanFM) [inline finetuner/tuner/pytorch/miner.py]: sorted is already sorted right?

Comment (bwanglzu) [reply finetuner/tuner/pytorch/miner.py]: refresh :)

File: tests/unit/tuner/torch/test_miner.py [hunk 2]
```diff
@@ -27,6 +42,11 @@ def labels():
     return [1, 3, 1, 3, 2, 4, 2, 4]
 
 
+@pytest.fixture
+def session_labels():
+    return [(1, 1), (2, -1), (2, 1), (1, -1), (3, -1), (3, 1), (3, -1), (3, -1)]
```
Comment (tadejsv) [inline tests/unit/tuner/torch/test_miner.py:47]: You need to take into account that each session has an anchor with label `0`

File: tests/unit/tuner/torch/test_miner.py [hunk 3]
```diff
@@ -75,3 +95,60 @@ def test_triplet_miner_given_insufficient_inputs(
...
+    for item in rv:
+        idx_left, idx_right, label = item
+        # find corresponded label idx
+        label_left = session_labels[idx_left]
+        label_right = session_labels[idx_right]
+        if label_left == label_right:
+            expected_label = 1
+        else:
+            expected_label = -1
+        assert label == expected_label
```
Comment (tadejsv) [inline tests/unit/tuner/torch/test_miner.py:114]: Here I would prefer if you explicitly list the tuples you expect to be produced (since it's only 8 of them, it's feasible) or produce them by a "reliable" method.

The way test is written now does not pin down the result enough in my opinion

File: tests/unit/tuner/torch/test_miner.py [hunk 4]
```diff
@@ -75,3 +95,60 @@ def test_triplet_miner_given_insufficient_inputs(
...
+def test_siamese_session_miner_given_insufficient_inputs(
+    embeddings, session_labels, siamese_session_miner, cut_index
+):
+    embeddings = embeddings[:cut_index]
+    session_labels = session_labels[:cut_index]
+    rv = list(siamese_session_miner.mine(embeddings, session_labels))
+    assert len(rv) == 0
+
+
+def test_triplet_session_miner(embeddings, session_labels, triplet_session_miner):
```
Comment (tadejsv) [inline tests/unit/tuner/torch/test_miner.py:127]: Same for this test, list expected result explicitly.

File: finetuner/tuner/pytorch/miner.py [hunk 5]
```diff
@@ -34,11 +49,61 @@ def mine(
...
+        sorted_labels_with_index = sorted(labels_with_index, key=lambda x: x[0])
+        for session, group in groupby(sorted_labels_with_index, lambda x: x[0]):
+            _, session_labels, session_indices = zip(*group)
+            pairs = _generate_all_possible_pairs(session_labels)
+            rv.extend(
+                [
+                    (session_indices[left_idx], session_indices[right_idx], label)
+                    for left_idx, right_idx, label in pairs
+                ]
+            )
```
Comment (tadejsv) [inline finetuner/tuner/pytorch/miner.py:78]: Please, this should be a for loop. Unless you can show we get significant time savings (not microseconds per batch), readability has absolute priority

File: finetuner/tuner/pytorch/miner.py [hunk 6]
```diff
@@ -34,11 +49,61 @@ def mine(
...
+            _, session_labels, session_indices = zip(*group)
+            triplets = _generate_all_possible_triplets(session_labels)
+            rv.extend(
+                [
+                    (
+                        session_indices[left_idx],
+                        session_indices[middle_idx],
+                        session_indices[right_idx],
+                    )
+                    for left_idx, middle_idx, right_idx in triplets
```
Comment (tadejsv) [inline finetuner/tuner/pytorch/miner.py:106]: Same - for loop unless you can demonstrate significant time savings

File: finetuner/tuner/pytorch/miner.py [hunk 7]
```diff
@@ -34,11 +49,61 @@ def mine(
...
+                    (session_indices[left_idx], session_indices[right_idx], label)
+                    for left_idx, right_idx, label in pairs
+                ]
+            )
+        return rv
+
+
+class TripletSessionMiner(BaseMiner):
+    def mine(
+        self, embeddings: List[AnyTensor], labels: List[Tuple[int, int]]
```
Comment (tadejsv) [inline finetuner/tuner/pytorch/miner.py:84]:
```suggestion
        self, embeddings: AnyTensor, labels: List[Tuple[int, int]]
```

File: finetuner/tuner/pytorch/miner.py [hunk 8]
```diff
@@ -34,11 +49,61 @@ def mine(
...
-        np.fill_diagonal(matches, 0)
-        triplets = np.expand_dims(matches, 2) * np.expand_dims(diffs, 1)
-        idxes_anchor, idxes_pos, idxes_neg = np.where(triplets)
-        return list(zip(idxes_anchor, idxes_pos, idxes_neg))
+        return _generate_all_possible_triplets(labels)
+
+
+class SiameseSessionMiner(BaseMiner):
+    def mine(
+        self, embeddings: List[AnyTensor], labels: List[Tuple[int, int]]
```
Comment (tadejsv) [inline finetuner/tuner/pytorch/miner.py:57]:
```suggestion
        self, embeddings: AnyTensor, labels: List[Tuple[int, int]]
```

It's not a list, simply a tensor with first dimension `N`

File: finetuner/tuner/pytorch/miner.py [hunk 9]
```diff
@@ -34,11 +49,61 @@ def mine(
...
+    ) -> List[Tuple[int, ...]]:
```
Comment (tadejsv) [inline finetuner/tuner/pytorch/miner.py:58]:
```suggestion
    ) -> List[Tuple[int, int, int]]:
```

File: finetuner/tuner/pytorch/miner.py [hunk 10]
```diff
@@ -34,11 +49,61 @@ def mine(
...
+                    for left_idx, right_idx, label in pairs
+                ]
+            )
+        return rv
+
+
+class TripletSessionMiner(BaseMiner):
+    def mine(
+        self, embeddings: List[AnyTensor], labels: List[Tuple[int, int]]
+    ) -> List[Tuple[int, ...]]:
```
Comment (tadejsv) [inline finetuner/tuner/pytorch/miner.py:85]:
```suggestion
    ) -> List[Tuple[int, int, int]]:
```

File: finetuner/tuner/pytorch/miner.py [hunk 11]
```diff
@@ -34,11 +49,61 @@ def mine(
...
+        """Generate triplets from input embeddings and labels.
+
+        :param embeddings: embeddings from model, should be a list of Tensor objects.
+        :param labels: labels of each embeddings, consist of session id and label.
+        :return: triplet of label indices follows the order of anchor, positive and negative.
+        """
+        assert len(embeddings) == len(labels)
+        rv = []
+        labels_with_index = [item + (index,) for index, item in enumerate(labels)]
+        sorted_labels_with_index = sorted(labels_with_index, key=lambda x: x[0])
```
Comment (tadejsv) [inline finetuner/tuner/pytorch/miner.py:95]: Can I ask why are we sorting? We are doing `groupby` later anyway

File: finetuner/tuner/pytorch/miner.py [hunk 12]
```diff
@@ -34,11 +49,61 @@ def mine(
...
+        for session, group in groupby(sorted_labels_with_index, lambda x: x[0]):
+            _, session_labels, session_indices = zip(*group)
+            triplets = _generate_all_possible_triplets(session_labels)
```
Comment (tadejsv) [inline finetuner/tuner/pytorch/miner.py:98]: This implementation is actually wrong. You can not assume that items with label `-1` are similar among each other. I can also see that from your test, because for session `3` you should also get 0 triplets as it is now, but you say you get 6

**Review by tadejsv — CHANGES_REQUESTED** [review]

File: finetuner/tuner/pytorch/miner.py (hunk 11 above)
Comment (bwanglzu) [reply finetuner/tuner/pytorch/miner.py:95]: if you [read](https://stackoverflow.com/a/8116674), groupby only works on sorted data, it's stupid but true.

File: finetuner/tuner/pytorch/miner.py (hunk 12 above)
Comment (bwanglzu) [reply finetuner/tuner/pytorch/miner.py:98]: i see, you're right

File: finetuner/tuner/pytorch/miner.py [hunk 13]
```diff
@@ -42,3 +42,71 @@ def mine(
...
+        rv = []
+        labels_with_index = [item + (index,) for index, item in enumerate(labels)]
+        sorted_labels_with_index = sorted(labels_with_index, key=lambda x: x[0])
+        for session, group in groupby(sorted_labels_with_index, lambda x: x[0]):
+            _, session_labels, session_indices = zip(*group)
+            session_pairs = []
+            for left, right in combinations(enumerate(session_labels), 2):
+                if left[1] >= 0 and right[1] >= 0:
+                    # 0 represents for anchor, 1 for positive, they form positive pairs.
+                    session_pairs.append((left[0], right[0], 1))
```
Comment (tadejsv) [inline finetuner/tuner/pytorch/miner.py:68]: I would append to `rv` here already, so we don't need to do another loop

File: finetuner/tuner/pytorch/miner.py [hunk 14]
```diff
@@ -42,3 +42,71 @@ def mine(
...
+          anchor, positive, negative are 0, 1 and -1.
+        :return: triplet of label indices follows the order of anchor, positive and negative.
+        """
+        assert len(embeddings) == len(labels)
+        rv = []
+        labels_with_index = [item + (index,) for index, item in enumerate(labels)]
+        sorted_labels_with_index = sorted(labels_with_index, key=lambda x: x[0])
+        for session, group in groupby(sorted_labels_with_index, lambda x: x[0]):
+            _, session_labels, session_indices = zip(*group)
+            for left, middle, right in permutations(enumerate(session_labels), 3):
```
Comment (tadejsv) [inline finetuner/tuner/pytorch/miner.py:101]: I'm thinking we can make this more efficient. Currently most of permutations we loop over will get discarded.

We already know that only positive and anchors can be at first two places, and negative has to be at the last place. So what you could do is to:
- first, get a list of positive/anchor indices
- get a list of negative indices
- loop over permutations of size 2 for pos/anchor
- inside this loop, loop over negative indices
- add each such combination to result

**Review by tadejsv — APPROVED** [review]
Looks good!

**Comment by bwanglzu** [comment]
note: i fixed tf version in this PR. this needs to be investigated in the future.
